File: algorithms/python/最小的k个数/最小的k个数.py
# Source : https://leetcode.cn/problems/zui-xiao-de-kge-shu-lcof/description/
# Author : Yanan Wang
# Date   : 2023-08-23

# 本题与《215. 数组中的第K个最大元素》非常相似，不同之处是本题要返回前k个数，
# 那么就只需要在分区之后，改变一下动作即可
#      -----------------------
#      ^          ^          ^ 
#     <key      ==key       >key  
#      a个        b个        c个
# 当 a < k时：去左边区域寻找k个最小元素
# 当 a + b <= k时：直接返回即可，因为此时已经找到了k个最小的数，只需在递归整体完成之后返回arr[:k]
# 当上面两个条件都不成立时：说明还有k-a-b个元素落在右边区域，那么就去右边区域寻找k-a-b个最小元素

class Solution:
    def getLeastNumbers(self, arr: List[int], k: int) -> List[int]:
        def quickSelect(arr, low, high, k):
            if low >= high:
                return
            key = arr[random.randint(low, high)]
            left, right = low, high
            i = low
            while i <= right:
                if arr[i] < key:
                    arr[left], arr[i] = arr[i], arr[left]
                    left += 1
                    i += 1
                elif arr[i] == key:
                    i += 1
                else:
                    arr[i], arr[right] = arr[right], arr[i]
                    right -= 1
            # [low, left)  [left, right]  (right, high]
            leftNums = left - low
            midNums = right - left + 1
            rightNums = high - right
            if leftNums > k:
                quickSelect(arr, low, left - 1, k)
            elif leftNums + midNums >= k:
                return
            else:
                quickSelect(arr, right + 1, high, k - leftNums - midNums)


        quickSelect(arr, 0, len(arr) - 1, k)
        return arr[:k]

File: algorithms/cpp/最小的k个数/最小的k个数.cpp
// Source : https://leetcode.cn/problems/zui-xiao-de-kge-shu-lcof/description/
// Author : Yanan Wang
// Date   : 2023-11-25

class Solution {
public:
    void QuickSelect(vector<int> &stock, int begin, int end, int cnt) {
        if (begin >= end) {
            return;
        }
        int key = stock[begin + (rand() % (end - begin + 1))];
        int i = begin;
        int left = begin - 1;
        int right = end + 1;
        while (i < right) {
            if (stock[i] < key) {
                swap(stock[++left], stock[i++]);
            } else if (stock[i] == key) {
                ++i;
            } else {
                swap(stock[i], stock[--right]);
            }
        }
        int left_nums = left - begin + 1;
        int mid_nums = right - left - 1;
        if (left_nums > cnt) {
            QuickSelect(stock, begin, left, cnt);
        } else if (left_nums + mid_nums >= cnt) {
            return;
        } else {
            QuickSelect(stock, right, end, cnt - left_nums - mid_nums);
        }
    }
    vector<int> inventoryManagement(vector<int>& stock, int cnt) {
        srand(time(nullptr));
        QuickSelect(stock, 0, stock.size() - 1, cnt);
        return {stock.begin(), stock.begin() + cnt};
    }
};